src/storage/src/decode/mod.rs

```diff
@@ -56,77 +53,88 @@ mod protobuf;
 ) -> (Collection<G, Row, Diff>, PressOnDropButton) {
     let channel_rx = Rc::new(RefCell::new(VecDeque::new()));
     let activator_set: Rc<RefCell<Option<SyncActivator>>> = Rc::new(RefCell::new(None));
 
-    let mut builder = AsyncOperatorBuilder::new("CDCv2-Decode".to_owned(), input.scope());
-
-    let mut input_handle = builder.new_disconnected_input(
-        &input.inner,
-        Exchange::new(|(x, _, _): &(SourceOutput<FromTime>, _, _)| x.key.hashed()),
-    );
-
+    let mut row_buf = Row::default();
+    let mut vector = vec![];
     let channel_tx = Rc::clone(&channel_rx);
     let activator_get = Rc::clone(&activator_set);
-    builder.build(move |_| async move {
-        let registry = match csr_connection {
-            None => None,
-            Some(conn) => Some(
-                // This also panics on connections errors. cdc_v2 is unused so we don't handle
-                // errors right now.
-                conn.connect(&storage_configuration)
-                    .await
-                    .expect("CSR connection unexpectedly missing secrets"),
-            ),
-        };
-
-        // We have already checked validity of the schema by now, so this can't fail.
-        let mut resolver =
-            ConfluentAvroResolver::new(&schema, registry, confluent_wire_format).unwrap();
-
-        while let Some(event) = input_handle.next().await {
-            let AsyncEvent::Data(_time, data) = event else {
-                continue;
-            };
-
-            for (data, _time, _diff) in data {
-                let value = match data.value.unpack_first() {
-                    Datum::Bytes(value) => value,
-                    Datum::Null => continue,
-                    _ => unreachable!("invalid datum"),
-                };
-                let (mut data, schema, _) = match resolver.resolve(&*value).await {
-                    Ok(Ok(ok)) => ok,
-                    // TODO: restart the dataflow on transient errors
-                    Ok(Err(e)) | Err(e) => {
-                        error!("Failed to get schema info for CDCv2 record: {}", e);
+    let pact = Exchange::new(|(x, _, _): &(DecodeResult<FromTime>, _, _)| x.key.hashed());
+    input.inner.sink(pact, "CDCv2Unpack", move |input| {
+        while let Some((_, data)) = input.next() {
+            data.swap(&mut vector);
+            // The inputs are rows containing two columns that encode an enum, i.e only one of them
+            // is ever set and the other one is set. This is the convention we follow in our Avro
+            // decoder. When the first field of the record is set then we have a data message.
+            // Otherwise we have a progress message.
+            for (row, _time, _diff) in vector.drain(..) {
+                let mut record = match &row.value {
+                    Some(Ok(row)) => row.iter(),
+                    Some(Err(err)) => {
+                        error!("Ignoring errored record: {err}");
```

Comment:
hmmm, we might need to rethink this someday

Reply:
yes, this only replicates previous behavior